### PR TITLE
Add compressed byte output and new CLI options for branching

### PR DESCRIPTION
**Pull request by dkpoult**
Adds the following features:
- Now shows the total number of compressed bytes read as part of the info log.
- Adds the ability to use regex in branch names using the `--branches` CLI argument. Includes another flag `--no-regex` to disable it.
- Adds an `--all-branches` flag to use all branches, essentially a shorthand for `--branches ".*"` with regex forced on.
- Adds extra testing for using regex branch names.

### CODE REVIEW

File: src/root_readspeed.cxx [hunk 1]
```diff
@@ -20,6 +20,7 @@ void PrintThroughput(const Result &r)
    std::cout << "Real time:\t\t\t" << r.fRealTime << " s\n";
    std::cout << "CPU time:\t\t\t" << r.fCpuTime << " s\n";
    std::cout << "Uncompressed data read:\t\t" << r.fUncompressedBytesRead << " bytes\n";
+   std::cout << "Compressed data read:\t\t" << r.fCompressedBytesRead << " bytes\n";
 
    std::cout << "Throughput:\t\t\t" << r.fUncompressedBytesRead / r.fRealTime / 1024 / 1024 << " MB/s\n";
```
Comment (eguiraud) [inline src/root_readspeed.cxx]: please change this to `"Uncompressed throughput" (or something else that indicates that this is uncompressed bytes/s) and also add a `"Compressed throughput"`. I think both numbers are interesting, the former is independent of the compression algorithm and it is useful to compare across compression algorithms and levels, the latter is close to the raw I/O bandwidth the application made use of.

File: src/root_readspeed.cxx [hunk 2]
```diff
@@ -70,14 +77,22 @@ Args ParseArgs(int argc, char **argv)
...
 int main(int argc, char **argv)
 {
-   const auto args = ParseArgs(argc, argv);
+   auto args = ParseArgs(argc, argv);
+
    if (!args.fShouldRun)
       return 1; // ParseArgs has printed the --help, has run the --test or has encountered an issue and logged about it
+   if (args.fAllBranches) {
+      if (!args.fData.fBranchNames.empty())
+         throw std::runtime_error("Can't use 'all-branches' argument and specify branches as well.");
```
Comment (eguiraud) [inline src/root_readspeed.cxx]: It might be clearer for the average user if we explicitly mention which options are conflicting: "ERROR: options `--all-branches` and `--branches` are mutually exclusive, please only use one".

File: src/ReadSpeed.hxx [hunk 3]
```diff
@@ -62,6 +65,56 @@ struct ByteData {
    ULong64_t fCompressedBytesRead;
 };
 
+// Gets the names of all branches in a tree, accounts for friends.
+// Copied straight from RDFInterfaceUtils
+void GetTopLevelBranchNamesImpl(TTree &t, std::set<std::string> &bNamesReg, ROOT::RDF::ColumnNames_t &bNames,
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: why do we need to copy the code? can't we just invoke `ROOT::Internal::RDF::GetTopLevelBranchNames(t)` after including `ROOT/RDF/InterfaceUtils.hxx`?

File: src/ReadSpeed.hxx [hunk 4]
```diff
@@ -52,9 +60,66 @@ struct EntryRange {
...
+   std::set<TTree *> analysedTrees;
+   GetTopLevelBranchNamesImpl(t, bNamesSet, bNames, analysedTrees);
+   return bNames;
+}
+
 // Read branches listed in branchNames in tree treeName in file fileName, return number of uncompressed bytes read.
-inline ULong64_t ReadTree(const std::string &treeName, const std::string &fileName,
-                          const std::vector<std::string> &branchNames, EntryRange range = {-1, -1})
+inline ByteData ReadTree(const std::string &treeName, const std::string &fileName,
+                          const std::vector<std::string> &branchRegexes, const bool &useRegex,
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]:
```suggestion
                          const std::vector<std::string> &branchRegexes, bool useRegex,
```

see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

File: src/ReadSpeed.hxx [hunk 5]
```diff
@@ -70,17 +135,31 @@ inline ULong64_t ReadTree(const std::string &treeName, const std::string &fileNa
...
+         return std::regex_match(thisName, branchRegex);
+      };
+
+      const auto iterator = std::find_if(branchRegexes.begin(), branchRegexes.end(), matchBranch);
+      return iterator != branchRegexes.end();
+   };
+   std::copy_if(unfilteredBranchNames.begin(), unfilteredBranchNames.end(), std::back_inserter(branchNames), filterBranchName);
+
+   std::vector<TBranch *> branches;
+   for (auto bName : branchNames) {
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: avoid string copies by taking a reference (to const, why not):

```suggestion
   for (const auto &bName : branchNames) {
```

File: test/tests.cpp [hunk 6]
```diff
@@ -8,15 +8,19 @@
...
    if (gSystem->AccessPathName(fname.c_str()) == false) // then the file already exists: weird return value convention
       return;                                           // nothing to do
 
    TFile f(fname.c_str(), "recreate");
    TTree t("t", "t");
-   int x = 42;
-   t.Branch("x", &x);
+
+   int var = 42;
+   for (const auto b : branchNames) {
```
Comment (eguiraud) [inline test/tests.cpp]: avoid string copies:

```suggestion
   for (const auto &b : branchNames) {
```

**Review by eguiraud — COMMENTED** [review]
Thank you @dkpoult ! Some comments below and inline:

### 1. Compilation issue

The current code for this PR does not compile on my laptop, with:
```
In file included from /home/blue/Scratchpad/work/root-readspeed/src/root_readspeed.cxx:4:
/home/blue/Scratchpad/work/root-readspeed/src/ReadSpeed.hxx:10:10: fatal error: ROOT/RDF/Utils.hxx: No such file or directory
   10 | #include <ROOT/RDF/Utils.hxx>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
Probably the CMakeLists.txt needs to be updated to mention the dependency on the ROOTDataFrame module, otherwise the corresponding include directories are not passed to the compiler invocation.

### 2. About regex-matching

It would be better if we could do the regex-matching outside of the hot loop rather than inside. We don't want to measure the extra time we spend compiling N regexes and matching them against (potentially thousands of) branch names. We can probably peek at the branches in the first tree and just do the matching against that. Varying schemas across input trees is not a very common use case.

Also, given the new logic, I am not sure whether we still error out correctly if a specified branch does not exist, e.g. if `--branches x` is passed and branch `x` does not exist we currently error out, but if `x` is now simply seen as a regex that doesn't match anything then it will silently do nothing.

Some ideas to improve the situation:
- always put the provided regexes between `^` and `$`, so that `x` will match exactly `^x$`, it will _not_ match `extra` and `fox`, and the user has to specify `.*x.*` if that's what they want
- have a dedicated `--branch-regexes` that's mutually exclusive with `--branches`, like many CLI tools have (e.g. [find](https://linux.die.net/man/1/find))
- should we communicate the list of matched branches in output? should we have a verbose mode for that?

### 3. General comments
- the fix to the `std::accumulate` call that is in the same commit as a test addition should be instead squashed with the commit that introduces the wrong `std::accumulate` call
- please pass the code through clang-format, there is some formatting to be fixed. you can do it with `clang-format -i <files>` or even better with `git clang-format --diff main branch-feature-changes | git apply` 
- please add a test that checks we compute the correct values for the compressed bytes

File: src/ReadSpeed.hxx (hunk 3 above)
Comment (dkpoult) [reply src/ReadSpeed.hxx]: I get an error that `ROOT::Internal::RDF::GetTopLevelBranchNames(*t)` is undefined when I try to build it (after including `ROOTDataFrame` in the `CMakeLists.txt` required components list as well), I assumed it's because it's only internally accessible but it's very possible I'm just doing something dumb

Comment (eguiraud) [reply src/ReadSpeed.hxx]: `Internal` is just a name, as long as the function prototype is available at compile time (through the right `#include` directive) and the associated symbol is available at link time (the build system needs to link the right library) things should work fine. i'll be in tomorrow, let's take a look together

File: src/ReadSpeed.hxx [hunk 7]
```diff
@@ -7,6 +7,7 @@
 #include <ROOT/TSeq.hxx>
 #include <ROOT/TThreadExecutor.hxx>
 #include <ROOT/TTreeProcessorMT.hxx> // for TTreeProcessorMT::GetTasksPerWorkerHint
+#include <ROOT/RDF/InterfaceUtils.hxx>
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]:
```suggestion
#include <ROOT/RDF/InterfaceUtils.hxx> // for ROOT::Internal::RDF::GetTopLevelBranchNames
```

File: src/ReadSpeed.hxx [hunk 8]
```diff
@@ -52,35 +59,88 @@ struct EntryRange {
...
+      const auto iterator = std::find_if(regexes.begin(), regexes.end(), matchBranch);
+      return iterator != regexes.end();
+   };
+   std::copy_if(unfilteredBranchNames.begin(), unfilteredBranchNames.end(), std::back_inserter(branchNames),
+                filterBranchName);
+
+   if (branchNames.empty() && useRegex)
+      throw std::runtime_error("Provided branch regexes didn't match any branches in the tree.");
+   if (usedRegexes.size() != regexes.size()) {
+      std::cout << "The following branches/regexes weren't found in the tree:" << std::endl;
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]:
```suggestion
      std::cerr << "The following branches/regexes weren't found in the tree:\n";
```

see also https://cppkid.wordpress.com/2008/08/27/why-i-prefer-n-to-stdendl/

File: src/ReadSpeed.hxx [hunk 9]
```diff
@@ -52,35 +59,88 @@ struct EntryRange {
...
+      for (const auto &regex : regexes) {
+         if (usedRegexes.find(regex) == usedRegexes.end())
+            std::cout << "\t" + regex << std::endl;
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: no need to create a `const char*` when we really just need a `char`:
```suggestion
            std::cerr << '\t' + regex << '\n';
```

File: src/ReadSpeed.hxx [hunk 10]
```diff
@@ -52,35 +59,88 @@ struct EntryRange {
...
+      }
+      throw std::runtime_error("Some branches/regexes weren't found in the tree...");
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: if we are throwing anyway, shouldn't the messages we print above just be part of the exception message?

File: src/ReadSpeed.hxx [hunk 11]
```diff
@@ -89,30 +149,41 @@ inline ULong64_t ReadTree(const std::string &treeName, const std::string &fileNa
...
+   ULong64_t compressedBytesRead = 0;
+
    TStopwatch sw;
-   sw.Start();
 
-   auto treeIdx = 0;
-   ULong64_t bytesRead = 0;
    for (const auto &fName : d.fFileNames) {
-      bytesRead += ReadTree(d.fTreeNames[treeIdx], fName, d.fBranchNames);
+      const auto branchNames = GetMatchingBranchNames(fName, d.fTreeNames[treeIdx], d.fBranchNames, d.fUseRegex);
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: `GetMatchingBranchNames` is a fairly expensive call. If the user did not specify regexes we should just use the old logic which is much faster. I suspect that most of the runtime of the test suite is spent in `GetMatchingBranchNames` at the moment.

File: src/root_readspeed.cxx [hunk 12]
```diff
@@ -45,16 +51,48 @@ Args ParseArgs(int argc, char **argv)
...
+            return {};
+         }
+         branchState = EBranchState::kRegular;
+      } else if (arg.compare("--branches-regex") == 0) {
          argState = EArgState::kBranches;
-      else if (std::strcmp(argv[i], "--threads") == 0)
+         if (branchState != EBranchState::kNone && branchState != EBranchState::kRegex) {
+            std::cerr << "Options --all-branches, --branches, and --branches-regex are mutually exclusive. You can use "
+                         "only one."
+                      << std::endl;
```
Comment (eguiraud) [inline src/root_readspeed.cxx]: could do:

```suggestion
            std::cerr << branchOptionsErrMsg;
```

after lifting the message into a std::string to avoid repeating the string three times.

**Review by eguiraud — CHANGES_REQUESTED** [review]
thank you @dkpoult ! a few minor things, but this looks almost ready to go

File: .gitignore [hunk 13]
```diff
@@ -2,3 +2,4 @@ tags
 build/
 *.swp
 *~
+.vscode
```
Comment (eguiraud) [inline .gitignore:5]: this belongs in a different commit

File: src/ReadSpeed.hxx [hunk 14]
```diff
@@ -52,35 +59,88 @@ struct EntryRange {
...
+      const auto iterator = std::find_if(regexes.begin(), regexes.end(), matchBranch);
+      return iterator != regexes.end();
+   };
+   std::copy_if(unfilteredBranchNames.begin(), unfilteredBranchNames.end(), std::back_inserter(branchNames),
+                filterBranchName);
+
+   if (branchNames.empty())
+      throw std::runtime_error("Provided branch regexes didn't match any branches in the tree.");
+   if (usedRegexes.size() != regexes.size()) {
+      std::stringstream errStream;
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: this is not performance-sensitive code, but just appending to a string is so much faster than using stringstream that it almost makes using stringstream here bad practice: https://www.quick-bench.com/q/yNqfgq9vKdAHKExFg-VnyFDZG-E

File: src/ReadSpeed.hxx [hunk 15]
```diff
@@ -52,35 +59,88 @@ struct EntryRange {
...
    std::unique_ptr<TTree> t(f->Get<TTree>(treeName.c_str()));
    if (t == nullptr)
       throw std::runtime_error("Could not retrieve tree '" + treeName + "' from file '" + fileName + '\'');
+
    t->SetBranchStatus("*", 0);
-   std::vector<TBranch *> branches(branchNames.size());
-   auto getBranch = [&t](const std::string &bName) {
+
+   std::vector<TBranch *> branches;
+   for (auto &bName : branchNames) {
```
Comment (eguiraud) [inline src/ReadSpeed.hxx]: why not const:

```suggestion
   for (const auto &bName : branchNames) {
```

**Review by eguiraud — CHANGES_REQUESTED** [review]
Looks good. As a `git` exercise we should split the `.gitignore` change from the last commit and move it to another commit, and we should split commit `Fix (un)compressed byte info summing` in one commit that does what the message says and another that improves the test suite.

File: src/ReadSpeed.hxx (hunk 14 above)
Comment (dkpoult) [reply src/ReadSpeed.hxx]: Ah TIL, in Java it's always recommended to use StringBuilder instead of regular appending for performance reasons, interesting that it's the opposite in CPP

**Comment by dkpoult** [comment]
@eguiraud did the extra perf changes and split the commits

**Review by eguiraud — APPROVED** [review]
LGTM!
